CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs

```diff
@@ -0,0 +1,437 @@
+        {
+            get
+            {
+                return _instance;
+            }
+        }
+
+        private CharacterObject _currentTroop;
+        private Hero _currentAddedHero;
+        private MapEvent _currentPlayerEvent;
+        private bool playerIsInBattle;
+        private bool isFilling;
+        private bool _isloaded;
+        private Dictionary<string, int> _cultureCounts = new Dictionary<string, int>();
+        private Dictionary<string, PartyAttribute> _partyAttributes = new Dictionary<string, PartyAttribute>();
+        
+        private List<PartyAttribute> _activePartyAttributes;
+        private Action<float> deltaTime;
+
+        public List<PartyAttribute> GetActiveInvolvedParties()
+        {
+            return _activePartyAttributes;
+        }
+        
+        public EventHandler<BattleAttributesArgs> NotifyBattlePartyObservers;
+        public override void  RegisterEvents()
+        {
+            //Game Saving Events
+            CampaignEvents.OnBeforeSaveEvent.AddNonSerializedListener(this, OnGameSaving());
+            CampaignEvents.OnGameLoadFinishedEvent.AddNonSerializedListener(this, OnGameLoaded);
+            CampaignEvents.OnNewGameCreatedPartialFollowUpEndEvent.AddNonSerializedListener(this,OnNewGameCreatedPartialFollowUpEnd);
+            
+            //Tick events
+            CampaignEvents.TickEvent.AddNonSerializedListener(this, deltaTime => FillWindsOfMagic(deltaTime));
+            CampaignEvents.DailyTickEvent.AddNonSerializedListener(this,OnDailyTick());
+            
+            //Events and Battles
+            CampaignEvents.MapEventStarted.AddNonSerializedListener(this,EventCreated);
+            CampaignEvents.BattleStarted.AddNonSerializedListener(this,OnBattleStarted);
+            CampaignEvents.BeforeMissionOpenedEvent.AddNonSerializedListener(this, OnMissionStarted);
+            
+            //Parties created and destroyed
+            CampaignEvents.MobilePartyCreated.AddNonSerializedListener(this,RegisterParty);
+            CampaignEvents.MobilePartyDestroyed.AddNonSerializedListener(this,DeregisterParty);
+            
+            //Units added
+            CampaignEvents.OnPartySizeChangedEvent.AddNonSerializedListener(this,OnPartySizeChanged);
+            CampaignEvents.NewCompanionAdded.AddNonSerializedListener(this,OnCompanionAdded);
+            CampaignEvents.OnTroopRecruitedEvent.AddNonSerializedListener(this,OnTroopRecruited);
+            CampaignEvents.OnUnitRecruitedEvent.AddNonSerializedListener(this, OnPlayerRecruitedTroop);
+                //missing : OnTroopRemoved
+        }
+
+        private void OnMissionEnded(MapEvent mapEvent)
+        {
+            if (playerIsInBattle)
+            {
+                playerIsInBattle = false;
+            }
+        }
+
+        
+
+        private void OnPlayerRecruitedTroop(CharacterObject troop, int amount)
+        {
+       //     TOWCommon.Say("Player Troop Recruited");
```

Comment:
Commented out debug messages should be removed before merging

Reply:
Besides some I use for later testing and improvements, I reduced most on unnecessary ones